### ISSUE TITLE
Identify whether the relabeling in sklearn will cause problem

### ISSUE THREAD

**Issue by yangarbiter** [ntucllab/libact#27]

**Comment by lazywei** [comment]
@yangarbiter could you elaborate on this issue?

**Comment by yangarbiter** [comment]
@lazywei I've update the details.
Thanks.

**Comment by lazywei** [comment]
Are you talking about the [`sklearn.preprocessing.LabelEncoder`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/preprocessing/label.py#L96-L111)? If so, they indeed did that through applying `np.unique` on the label vector.

How about moving the missing labels to the end of the label vector and maintaining an "internal index" in our model?
```
def get_internal_idx(y):
    n_samples = y.shape[0]

    nan_idx = np.argwhere(np.isnan(y))[:, 0]
    ttl_idx = list(range(n_samples))

    s = set(nan_idx)
    return [_idx for _idx in ttl_idx if _idx not in s] + nan_idx.tolist()


y = np.array([np.nan, 0, 1, np.nan, 1])

intl_idx = get_internal_idx(y)
print(intl_idx)

intl_y = y[intl_idx]
print(intl_y)
```

The checking `return [_idx for ...]` is order-preserved. And if performance is a concern, this checking is better than `O(m * n)` where m and n is the number of non-nan and nan elements.

We can get this `intl_idx` for the first time training. And for the future updating, we always stick to this index. Then, we are guranteed to have the same "label order".

How do you think?

**Comment by yangarbiter** [comment]
I think maintaining an "internal index" and decode it before it outputs from our model should be enough (we can swap the row to the right order before return from predict_real). And simply add a new label to internal label if new label comes ( #9 ).

Though if we want to move the missing labels to the end of the label, we might have to also consider this issue  #9 . Since if we do so, it might be harder to extend to the situation which labeled pool contain only a subset of all possible labels.
